codegen/masm/src/codegen.rs

```diff
@@ -114,15 +200,41 @@ fn quadratic_element_square(writer: &mut Writer, n: u32) {
 
 impl<'ast> CodeGenerator<'ast> {
     pub fn new(ir: &'ast AirIR, config: CodegenConfig) -> CodeGenerator<'ast> {
+        // remove duplicates and sort period lengths in descending order, since larger periods will
+        // give smaller number of cycles (which means a smaller number of exponentiations)
+        let mut periods: Vec<usize> = ir.periodic_columns().iter().map(|e| e.len()).collect();
```

Comment:
Initially some of this code was in the visit methods. But that was starting to create implicit required ordering to the visited fields. To avoid having bugs, I moved the simple code to the constructor, and left the visitor for the iteration that uses the algebraic graph.